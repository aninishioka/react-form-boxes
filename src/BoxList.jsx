import { useState } from "react";
import { v4 as uuid } from 'uuid';
import NewBoxForm from "./NewBoxForm";
import Box from "./Box";

/** Render all Box components + NewBoxForm
 *
 * props: none
 *
 * state:
 * - boxes: [{width, height, color}]
 *
 * App -> BoxList -> Box, NewBoxForm
*/

function BoxList() {
    const [boxes, setBoxes] = useState([])

    function renderBoxes() {
        return (
            <ul>
                {boxes.map(box => (
                    <Box
                        key={box.id}
                        width={box.width}
                        height={box.height}
                        color={box.color}
                        removeSelf={() => removeBox(box.id)}
                    />
                ))}
            </ul>
        )
    }

    /** Update state with new box */
    function addBox(box) {
        const newBox = {...box, id: uuid()}
        setBoxes(boxes => [...boxes, newBox]);
    }

    /** Remove box from state */
    function removeBox(id) {
        setBoxes(boxes => boxes.filter(b => b.id !== id));
    }

    return (
        <div>
            <NewBoxForm addBox={addBox} />
            {renderBoxes()}
        </div>
    )
}

export default BoxList;