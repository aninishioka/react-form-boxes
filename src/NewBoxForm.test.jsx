import { it, expect } from "vitest";

import { render, fireEvent } from "@testing-library/react";

import BoxList from "./BoxList";


it("renders without crashing", function () {
  render(<BoxList />);
});


it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});


it("can add a new box", function () {
  const { container, getByLabelText, queryByText, getElementsByClassName} = render(<BoxList />);
  console.log(container.getElementsByClassName("Box"))

  // no items yet
  expect(container.getElementsByClassName("Box")[0] || null).not.toBeInTheDocument();

  const heightInput = getByLabelText("Height:");
  const widthInput = getByLabelText("Width:");
  const colorInput = getByLabelText("Background color:");
  const submitBtn = queryByText("Add new box!");

  // fill out the form
  fireEvent.change(heightInput, { target: { value: 100 } });
  fireEvent.change(widthInput, { target: { value: 100 } });
  fireEvent.change(colorInput, { target: { value: "blue" } });
  fireEvent.click(submitBtn);

  // item exists!
  expect(container.getElementsByClassName("Box")[0]).toBeInTheDocument();

});