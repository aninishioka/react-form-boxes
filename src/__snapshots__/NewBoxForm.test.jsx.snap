// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value="0"
      />
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value="0"
      />
      <label
        for="color"
      >
        Background color:
      </label>
      <input
        id="color"
        name="color"
        value=""
      />
      <button>
        Add new box!
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
